### PR TITLE
Fix magnetosphere project tapering when running with an ionosphere

### PR DESCRIPTION
**Pull request by markusbattarbee**
Fixes the magnetosphere project tapering when running with an ionosphere. (Old method always went for conductingsphere parameters which were set to default values)

Implements a new method for checking existence of boundary conditions, and also adds an ionospheric testpackage test.
In Vlasiator.cpp there's a renaming of the wrapper object for sysBoundaries as I don't like having a class and an instance of that class with basically the same name.

### CODE REVIEW

File: projects/Magnetosphere/Magnetosphere.h [hunk 1]
```diff
@@ -117,5 +118,7 @@ namespace projects {
    }; // class Magnetosphere
 } // namespace projects
 
+extern SysBoundary sysBoundaryWrapper;
```
Comment (ursg) [inline projects/Magnetosphere/Magnetosphere.h:121]: I don't quite see yet why this separate object is necessary. Is it just so that the existence of boundaries can be tested through a global object somewhere?

Renaming all the sysboundary accesses further down in vlasiator.cpp will probably again lead to conflicts with the time varying inflow branch (although they won't be hard to correct)

File: sysboundary/sysboundary.cpp [hunk 2]
```diff
@@ -343,6 +343,25 @@ bool SysBoundary::initSysBoundaries(Project& project, creal& t) {
    return success;
 }
 
+/*!\brief Boolean check if queried sysboundarycondition exists
+ * Note: this queries against the parsed list of names, not against the actual existing boundaries. 
+ * This is so that the project configuration (which is handled before grid and boundary initialization)
+ * can use this call.
```
Comment (ursg) [inline sysboundary/sysboundary.cpp:349]: Big thumbs up for an actually helpful and constructive comment here!

File: vlasiator.cpp [hunk 3]
```diff
@@ -71,6 +71,7 @@ void fpehandler(int sig_num)
 
 Logger logFile, diagnostic;
 static dccrg::Dccrg<SpatialCell,dccrg::Cartesian_Geometry> mpiGrid;
+SysBoundary sysBoundaryWrapper;
```
Comment (ursg) [inline vlasiator.cpp:74]: Instead of placing this here as a global variable, another option would be to move it into the objectWrapper. That way, we are not further polluting the global variable namespace and have a little bit of a better encapsulation of global simulation state.

File: fieldsolver/gridGlue.cpp [hunk 4]
```diff
@@ -161,12 +161,7 @@ void filterMoments(dccrg::Dccrg<SpatialCell,dccrg::Cartesian_Geometry>& mpiGrid,
                int refLevel = technicalGrid.get(i, j, k)->refLevel;
 
                // Skip pass
-               if (blurPass >= P::numPasses.at(refLevel) ||
-                  technicalGrid.get(i, j, k)->sysBoundaryFlag == sysboundarytype::DO_NOT_COMPUTE ||
-                  (technicalGrid.get(i, j, k)->sysBoundaryFlag != sysboundarytype::NOT_SYSBOUNDARY &&
-                  technicalGrid.get(i, j, k)->sysBoundaryLayer >= 2)
-               )
-               {
+               if (blurPass >= P::numPasses.at(refLevel) || technicalGrid.get(i, j, k)->sysBoundaryFlag != sysboundarytype::NOT_SYSBOUNDARY) {
```
Comment (ursg) [inline fieldsolver/gridGlue.cpp:164]: This is basically PR #669, isn't it?

**Review by ursg — APPROVED** [review]
Good fixes! My critical comments here in the review probably shouldn't block this from being merged. So let's go for it!

File: fieldsolver/gridGlue.cpp (hunk 4 above)
Comment (markusbattarbee) [reply fieldsolver/gridGlue.cpp:164]: Yes, I pulled that commit directly into this branch/PR thinking it'd neatly ignore it here, but that isn't the way of github it seems.

File: projects/Magnetosphere/Magnetosphere.h (hunk 1 above)
Comment (markusbattarbee) [reply projects/Magnetosphere/Magnetosphere.h:121]: Yep, I think passing sysboundaries down to all projects doesn't make much sense for this special case, so a global object was my preferred choice.
The naming may indeed cause some clash, but because it's been just a local object and not part of an interface, it should be minor (and I'd say helpful).

File: vlasiator.cpp (hunk 3 above)
Comment (markusbattarbee) [reply vlasiator.cpp:74]: That would indeed be an alternative option. I'l look into it.

**Comment by markusbattarbee** [comment]
So now it's better. Should work as intended everywhere. 

Had a look at plots of the various magnetospheric tests and they seem to be somewhat sane. 

I changed the parametrization of the 3D tests (level 0 cell size is now 30e6 m, instead of 40e6 for ionosphere or 27,2727272...e6 for Magnetosphere_3D_small, and extended the tails a bit.

Also now the refinement level check for boundaries is less strict (nearest neighbours should be enough) so we don't have tons of error messages in testpackage runs.

Diffs against dev are zero / testpackage numeric level everywhere except for Ionosphere where it's a tapering fix, of course. How about now @ursg ?

**Review by ykempf — COMMENTED** [review]
Also, do we want to make use of non-uniform temperatures with tapering to make the test case less homogeneous and more sensitive to stuff?

File: sysboundary/ionosphere.cpp [hunk 5]
```diff
@@ -2978,11 +2978,14 @@ namespace SBC {
         Readparameters::get(pop + "_ionosphere.VZ0", sP.V0[2]);
         Readparameters::get(pop + "_ionosphere.T", sP.T);
 
+        // Failsafe, if density or temperature is zero, read from Magnetosphere
+        // (compare the corresponding verbose handling in projects/Magnetosphere/Magnetosphere.cpp)
```
[the remark this answers is not in the document]
Comment (markusbattarbee) [reply sysboundary/ionosphere.cpp:2982]: The warning is printed in Magnetosphere.cpp, that's why it now says there's *verbose* handling there :)

File: sysboundary/conductingsphere.cpp [hunk 6]
```diff
@@ -100,7 +100,16 @@ namespace SBC {
         Readparameters::get(pop + "_Magnetosphere.nSpaceSamples", sP.nSpaceSamples);
         Readparameters::get(pop + "_Magnetosphere.nVelocitySamples", sP.nVelocitySamples);
 
-        speciesParams.push_back(sP);
+        // Failsafe, if density or temperature is zero, read from Magnetosphere
+        // (compare the corresponding verbose handling in projects/Magnetosphere/Magnetosphere.cpp)
```
[the remark this answers is not in the document]
Comment (markusbattarbee) [reply sysboundary/conductingsphere.cpp:104]: and same, done in Magnetosphere.cpp

File: testpackage/run_tests.sh [hunk 7]
```diff
@@ -136,31 +136,35 @@ do
...
                 absoluteValue=$($run_command_tools $diffbin --meshname=fsgrid  ${result_dir}/${comparison_vlsv[$run]} ${vlsv_dir}/${comparison_vlsv[$run]} ${variables[$i]} ${indices[$i]} |grep "The absolute 0-distance between both datasets" |gawk '{print $8}'  )
-#print the results      
+#print the results
+                echo "${variables[$i]}_${indices[$i]}                $absoluteValue                 $relativeValue    "
+
+            elif [[ "${variables[$i]}" == "ig_"* ]]
+            then
+                relativeValue=$($run_command_tools $diffbin --meshname=ionosphere  ${result_dir}/${comparison_vlsv[$run]} ${vlsv_dir}/${comparison_vlsv[$run]} ${variables[$i]} ${indices[$i]} |grep "The relative 0-distance between both datasets" |gawk '{print $8}'  )
+                absoluteValue=$($run_command_tools $diffbin --meshname=ionosphere  ${result_dir}/${comparison_vlsv[$run]} ${vlsv_dir}/${comparison_vlsv[$run]} ${variables[$i]} ${indices[$i]} |grep "The absolute 0-distance between both datasets" |gawk '{print $8}'  )
```
[the remark this answers is not in the document]
Comment (markusbattarbee) [reply testpackage/run_tests.sh:149]: Indeed no, awaiting the vlsvdiff update.

File: projects/Magnetosphere/Magnetosphere.cpp [hunk 8]
```diff
@@ -117,34 +118,28 @@ namespace projects {
...
-      } else {
+      } else if (sysBoundaryContainer.existSysBoundary("Ionosphere")) {
+         RP::get("ionosphere.radius", this->ionosphereRadius);
+         RP::get("ionosphere.centerX", this->center[0]);
+         RP::get("ionosphere.centerY", this->center[1]);
+         RP::get("ionosphere.centerZ", this->center[2]);
          RP::get("ionosphere.geometry", this->ionosphereGeometry);
+      } else {
+         if(myRank == MASTER_RANK) {
+            std::cerr<<"Warning in initializing Magnetosphere: Could not find inner boundary geometry!"<<std::endl;
```
[the remark this answers is not in the document]
Comment (markusbattarbee) [reply projects/Magnetosphere/Magnetosphere.cpp:137]: Ok, updated message

**Comment by markusbattarbee** [comment]
> Also, do we want to make use of non-uniform temperatures with tapering to make the test case less homogeneous and more sensitive to stuff?

We could indeed. I put this in the ionospheric one:
```
T = 0.8e6
rho  = 2.0e6
```
and could well put some similar small increases in the other ones if we want?

**Comment by ykempf** [comment]
> and could well put some similar small increases in the other ones if we want?

If we want to catch more readily issues, the more gradients the merrier...

**Comment by ursg** [comment]
Code looks good to me now. Feel free to merge! (Or still change the tests and then merge, at your own discretion.)
